File: deamhome/src/main/java/sosteam/deamhome/domain/account/entity/Account.java
package sosteam.deamhome.domain.account.entity;

import jakarta.validation.constraints.NotNull;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.DBRef;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;
import sosteam.deamhome.domain.faq.entity.Faq;
import sosteam.deamhome.domain.item.entity.Wishlist;
import sosteam.deamhome.domain.review.entity.Review;
import sosteam.deamhome.global.attribute.Role;
import sosteam.deamhome.global.attribute.SNS;
import sosteam.deamhome.global.entity.LogEntity;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

@Document
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Account extends LogEntity implements UserDetails {

    @NotNull
    @Indexed(unique = true)
    private String userId;

    @NotNull
    private LocalDateTime birthday;

    @NotNull
    private boolean sex;

    @NotNull
    private String zipcode;

    @NotNull
    private String address1;

    @NotNull
    private String address2;

    @NotNull
    private String address3;

    @NotNull
    private String address4;

    @NotNull
    private String pwd;

    @NotNull
    @Indexed(unique = true)
    private String email;

    @NotNull
    private boolean receiveEmail;

    @NotNull
    private String createdIP;

    @NotNull
    private String adminTxt;

    @NotNull
    private String snsId;

    @NotNull
    private SNS sns = SNS.NORMAL;

    @NotNull
    @Indexed(unique = true)
    private String phone;

    @NotNull
    @Indexed(unique = true)
    private String userName;

    private int point = 0;

    @NotNull
    private Role role = Role.ROLE_GUEST;

    @DBRef(lazy = true)
    @Setter
    private AccountStatus status;

    @DBRef(lazy = true)
    private List<Faq> faq;

    @DBRef(lazy = true)
    @Setter
    private Wishlist wishlist;

    @DBRef(lazy = true)
    private List<Review> reviews;

    public List<Faq> addFaq(Faq faq) {
        this.faq.add(faq);

        return this.faq;
    }

    public List<Review> addReview(Review review) {
        reviews.add(review);

        return this.reviews;
    }

    @Override
    public Collection<? extends GrantedAuthority> getAuthorities() {
        ArrayList<SimpleGrantedAuthority> simpleGrantedAuthorities = new ArrayList<>();
        simpleGrantedAuthorities.add(new SimpleGrantedAuthority(role.name()));
        return simpleGrantedAuthorities;
    }

    @Override
    public String getPassword() {
        return this.pwd;
    }

    @Override
    public String getUsername() {
        return this.userId;
    }

    @Override
    public boolean isAccountNonExpired() {
        return true;
    }

    @Override
    public boolean isAccountNonLocked() {
        return true;
    }

    @Override
    public boolean isCredentialsNonExpired() {
        return true;
    }

    @Override
    public boolean isEnabled() {
        return true;
    }
}

File: src/main/kotlin/sosteam/deamhome/domain/account/entity/Account.kt
package sosteam.deamhome.domain.account.entity

import com.fasterxml.jackson.annotation.JsonIgnore
import org.springframework.data.annotation.Id
import org.springframework.data.relational.core.mapping.Column
import org.springframework.data.relational.core.mapping.Table
import org.springframework.security.core.authority.SimpleGrantedAuthority
import sosteam.deamhome.domain.question.entity.Question
import sosteam.deamhome.global.attribute.Role
import sosteam.deamhome.global.attribute.SNS
import sosteam.deamhome.global.entity.LogEntity
import java.time.LocalDateTime

@Table("account")
data class Account(
	@Id
	var id: Long?,
	// unique column
	@Column("user_id")
	val userId: String,
	var pwd: String,
	val sex: Boolean,
	var birthday: LocalDateTime,
	var zipcode: String,
	var address1: String,
	var address2: String?,
	var address3: String?,
	var address4: String?,
	
	// unique column
	var email: String,
	@Column("receive_mail")
	var receiveMail: Boolean,
	
	@Column("created_ip")
	val createdIp: String,
	@Column("admin_txt")
	var adminTxt: String = "",
	@Column("sns_id")
	val snsId: String? = null,
	val sns: SNS = SNS.NORMAL,
	
	// unique column
	var phone: String?,
	
	// unique column
	@Column("user_name")
	var userName: String,
	
	var point: Int = 0,
	
	var role: Role = Role.ROLE_GUEST,
	
	@Column("login_at")
	var loginAt: LocalDateTime,
	
	) : LogEntity() {
	companion object {
		const val maxWishListSize = 100
	}
	
	private var questions: ArrayList<String> = ArrayList()
	
	private var wishlist: ArrayList<String> = ArrayList()
	
	private var reviews: ArrayList<Long?> = ArrayList()
	
	@Column("review_report_ban_logs")
	private var reviewReportBanLogs: ArrayList<LocalDateTime> = ArrayList()
	
	@Column("review_report_logs")
	private var reviewReportLogs: ArrayList<LocalDateTime> = ArrayList()
	
	fun getWishList(): List<String> {
		return wishlist.toList()
	}
	
	fun addWishListItem(itemId: String): List<String> {
		wishlist.add(itemId)
		return wishlist
	}
	
	fun removeWishListItem(itemId: String): List<String> {
		wishlist.remove(itemId)
		return wishlist
	}
	
	fun isItemIdInWishlist(itemId: String): Boolean {
		return wishlist.contains(itemId)
	}
	
	fun getWishListSize(): Int {
		return wishlist.size
	}
	
	fun addQuestion(question: Question): List<String> {
//		questions.add(question.id)
		return questions
	}
	
	fun addReview(review: Long?): List<Long?> {
		reviews.add(review)
		return reviews
	}
	
	fun removeReview(review: Long?): List<Long?> {
		reviews.remove(review)
		return reviews
	}
	
	fun addReviewReportLog(report: LocalDateTime): List<LocalDateTime> {
		reviewReportLogs.add(report)
		return reviewReportLogs
	}
	
	fun getReviewReportLogs(): List<LocalDateTime> {
		return reviewReportLogs
	}
	
	fun addReviewReportBanLog(report: LocalDateTime): List<LocalDateTime> {
		reviewReportBanLogs.add(report)
		return reviewReportBanLogs
	}
	
	fun getReviewReportBanLogs(): List<LocalDateTime> {
		return reviewReportBanLogs
	}
	
	@JsonIgnore
	fun getAuthorities(): List<SimpleGrantedAuthority> {
		val simpleGrantedAuthorities = ArrayList<SimpleGrantedAuthority>()
		simpleGrantedAuthorities.add(SimpleGrantedAuthority(role.name))
		return simpleGrantedAuthorities
	}
}